packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+    // required assetHolders
+    assetHolders.forEach(assetHolder => {
+      if (this.assetHolders.has(assetHolder.contractAddress)) {
+        return;
+      }
+      // TODO: Use the AssetHolder abi?
+      const contract = new Contract(
+        assetHolder.contractAddress,
+        ContractArtifacts.EthAssetHolderArtifact.abi,
+        this.provider
+      );
+      this._registerAssetHolderCallbacks(contract);
+      this.assetHolders.set(assetHolder.contractAddress, {...assetHolder, contract});
+    });
+
+    // Add the channel to service storage
+    this.store.set(channelId, {assetHolders, events: {}});
+
+    return Promise.resolve(undefined);
+  }
+
+  public async submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse> {
+    if (!this.channelWallet) {
+      throw new Error(
+        'Must attach channel wallet before registering channel, this will go away in v1'
+      );
+    }
+
+    // If the channel is not registered, do not send transactions
+    if (!this.store.has(channelId)) {
+      throw new Error(`Channel is not registered ${channelId}`);
+    }
+
+    // Create the errors object
+    const errors: {[k: number]: string} = {};
+    for (let attempt = 1; attempt < this.retries; attempt += 1) {
+      try {
+        const response = await this.queue.add(() => this._sendTransaction(tx));
+        return response;
+      } catch (e) {
+        // Store error
+        errors[attempt] = e.message;
+
+        // Only retry if it is a known error
+        const knownErr = KNOWN_ERRORS.find(err => e.message.includes(err));
+        if (!knownErr) {
+          throw new Error(
+            `Transaction failed to send with unknown error on attempt ${attempt}/5: ${JSON.stringify(
+              errors
+            )}`
+          );
+        }
+      }
+    }
+    throw new Error(`Unable to send transaction after 5 retries: ${JSON.stringify(errors)}`);
+  }
+
+  private async _sendTransaction(tx: MinimalTransaction): Promise<providers.TransactionResponse> {
+    // Get the onchain record of wallet nonce
+    const chainNonce = await this.wallet.getTransactionCount();
+
+    // Use higher of chain or memory nonce
+    const nonce = this.memoryNonce > chainNonce ? this.memoryNonce : chainNonce;
+
+    // TODO: should we set gas minimums or preferences?
+
+    // Send the transaction
+    const response = await this.wallet.sendTransaction(tx);
+    if (!response.hash) {
+      throw new Error(NO_TX_HASH);
+    }
+
+    // Update the memory nonce after sent to mempool
+    this.memoryNonce = nonce + 1;
+    return response;
+  }
+
+  // NOTE: could also do this in the `wait` callback in `submitTransaction`
+  // but would need more information about what event to parse from the
+  // receipt via the API
+  private _registerAssetHolderCallbacks(assetHolder: Contract): void {
+    assetHolder.on(
+      'Deposited',
```

Comment:
> We also have helper functions inside of @statechannels/nitro-protocol to filter these events I think.

I'll double check but it seemed like those helpers were for parsing events from transaction logs. 

> filtering API to restrict events to specific subsets of emitted data based on the contained fields

I'll try to add the restricted event filters, but my concern is around the filter dynamically updating when I register a new channel. I don't want to turn the listeners off and on to update the filter (seems like it could get weirdly racy) every time I add a new channel, and I don't want a registered channel to have a 1:1 relationship with listeners on the contract. This seems like something that I'm going to need tests to clear up though, but if we can do it with the ethers filter easily that would definitely be better

Reply:
Check out the EVT changes I made, there are filters I included at that level, each asset holder only has the event registered once.

We can try to add the ethers filters, but I think the `pipe` functions have a nicer API